server/sonar-server/src/main/java/org/sonar/server/qualityprofile/DefinedQProfileRepositoryImpl.java

```diff
@@ -0,0 +1,200 @@
+
+    return qProfilesByLanguage;
+  }
+
+  /**
+   * @return profiles by language
+   */
+  private ListMultimap<String, RulesProfile> buildRulesProfilesByLanguage() {
+    ListMultimap<String, RulesProfile> byLang = ArrayListMultimap.create();
+    for (ProfileDefinition definition : definitions) {
+      ValidationMessages validation = ValidationMessages.create();
+      RulesProfile profile = definition.createProfile(validation);
+      validation.log(LOGGER);
+      if (profile != null && !validation.hasErrors()) {
+        checkArgument(isNotEmpty(profile.getName()), "Profile created by Definition %s can't have a blank name", definition);
+        byLang.put(lowerCase(profile.getLanguage(), Locale.ENGLISH), profile);
+      }
+    }
+    return byLang;
+  }
+
+  private void validateAndClean(ListMultimap<String, RulesProfile> byLang) {
+    byLang.asMap().entrySet()
+      .removeIf(entry -> {
+        String language = entry.getKey();
+        if (languages.get(language) == null) {
+          LOGGER.info("Language {} is not installed, related Quality profiles are ignored", language);
+          return true;
+        }
+        Collection<RulesProfile> profiles = entry.getValue();
+        if (profiles.isEmpty()) {
+          LOGGER.warn("No Quality profiles defined for language: {}", language);
+          return true;
+        }
+        return false;
+      });
+  }
+
+  private Map<String, List<DefinedQProfile>> toQualityProfilesByLanguage(ListMultimap<String, RulesProfile> rulesProfilesByLanguage) {
+    Map<String, List<DefinedQProfile.Builder>> buildersByLanguage = Multimaps.asMap(rulesProfilesByLanguage)
+      .entrySet()
+      .stream()
+      .collect(Collectors.uniqueIndex(Map.Entry::getKey, DefinedQProfileRepositoryImpl::toQualityProfileBuilders));
+    return buildersByLanguage
+      .entrySet()
+      .stream()
+      .filter(DefinedQProfileRepositoryImpl::ensureAtMostOneDeclaredDefault)
+      .collect(Collectors.uniqueIndex(Map.Entry::getKey, entry -> toQualityProfiles(entry.getValue()), buildersByLanguage.size()));
+  }
+
+  /**
+   * Creates {@link DefinedQProfile.Builder} for each unique quality profile name for a given language.
+   * Builders will have the following properties populated:
+   * <ul>
+   *   <li>{@link DefinedQProfile.Builder#language language}: key of the method's parameter</li>
+   *   <li>{@link DefinedQProfile.Builder#name name}: {@link RulesProfile#getName()}</li>
+   *   <li>{@link DefinedQProfile.Builder#declaredDefault declaredDefault}: {@code true} if at least one RulesProfile
+   *       with a given name has {@link RulesProfile#getDefaultProfile()} is {@code true}</li>
+   *   <li>{@link DefinedQProfile.Builder#activeRules activeRules}: the concatenate of the active rules of all
+   *       RulesProfile with a given name</li>
+   * </ul>
+   */
+  private static List<DefinedQProfile.Builder> toQualityProfileBuilders(Map.Entry<String, List<RulesProfile>> rulesProfilesByLanguage) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this unused private "toQualityProfileBuilders" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)